api/src/main/java/org/apache/iceberg/types/Type.java

```diff
@@ -111,6 +112,23 @@ public PrimitiveType asPrimitiveType() {
     Object writeReplace() throws ObjectStreamException {
       return new PrimitiveHolder(toString());
     }
+
+    @Override
+    public int hashCode() {
+      return Objects.hashCode(typeId());
+    }
+
+    @Override
+    public boolean equals(Object other) {
```

Comment:
Need more eyes on this. The way we currently compare structs breaks `Schema` and `PartitionSpec` Kryo  serialization. We rely on a proxy in `writeReplace` but Kryo ignores that.

If we decide to go this way, I can submit this in a separate PR if we want to handle it separately.

Reply:
Okay, so Kryo bypasses the singleton logic that we have in the proxy. Is there another way around it? I'm concerned about places that don't even call `equals` and assume that identity is enough, like `someType == LongType.get()`.

Reply:
I think the implementation here is okay, but we should have thorough tests for this because we don't want to break equality. Probably best to do this in a separate PR with the tests.

Reply:
I am also concerned about this.

I think we have the following alternative options:
- Pass spec and schema as json strings and deserialize as needed.
- Register a custom Kryo serializer for needed classes.

I think we would want to avoid the second option as much as we can. Passing schema and spec as JSON could work. Thoughts, @rdblue?

Reply:
Yeah, that sounds like a good option. I think we should probably do both. There isn't too much overhead to `equals` so we should start ensuring we correctly use it all the time.

Reply:
Addressed in a separate PR.